### PR TITLE
Add PLaMo 2.1 VL support

### PR DESCRIPTION
**Pull request by mitmul**
This PR adds the support for [PLaMo 2.1 VL](https://huggingface.co/pfnet/plamo-2.1-8b-vl) model. This model was developed by a Japanese company and utilizes PLaMo, an LLM particularly adept at handling Japanese, as the language model in its VLM architecture.

## Summary

- Add native `plamo2vl` support, including the model, language, vision, and processor implementations needed to run Plamo 2.1 VL in `mlx_vlm`.
- Improve Plamo2VL compatibility across loading and generation by preserving special token IDs, routing prompts through the model-specific processor/chat template, returning `InputEmbeddingsFeatures` from `get_input_embeddings`, and disabling chunked prefill for this architecture.
- Extend the shared LLaVA vision stack to support `gelu_pytorch_tanh` / `gelu_new`, which is required by the Plamo2VL SigLIP vision encoder.
- Refactor the quantization class predicate into an importable `mlx_vlm.utils.get_class_predicate` helper so it can be imported cleanly by the test suite when running `python -m unittest discover tests/` as described in `CONTRIBUTING.md`.
- Update `CONTRIBUTING.md` to point the unittest command at the `mlx_vlm/` directory.

## Added Tests

- `mlx_vlm/tests/test_models.py`
  - Add `test_plamo2vl_vision` to verify that the new Plamo2VL vision stack produces the expected feature shape and correctly propagates the `gelu_pytorch_tanh` activation setting.
  - Add `test_plamo2vl_model_config_preserves_special_token_ids` to ensure `ModelConfig.from_dict()` preserves `bos_token_id`, `eos_token_id`, and `pad_token_id` from model configs.
  - Add `test_plamo2vl_get_input_embeddings_accepts_mask` to verify that `get_input_embeddings()` accepts caller-provided mask kwargs and returns `InputEmbeddingsFeatures` with the expected embedding shape.
- `mlx_vlm/tests/test_prompt_utils.py`
  - Add `test_apply_chat_template_uses_plamo2vl_processor` to ensure `plamo2vl` prompt formatting is delegated to the model-specific processor instead of the generic chat-template path.
- `mlx_vlm/tests/test_utils.py`
  - Add `test_load_processor_prefers_custom_processor` to verify that `load_processor()` prefers a model-defined `Processor.from_pretrained()` implementation over `AutoProcessor.from_pretrained()` and still attaches the detokenizer correctly.
- `mlx_vlm/tests/test_generate.py`
  - Extend `test_generate_cli_smoke` to verify that the CLI forwards the original prompt list unchanged, passes `num_images=1` into `apply_chat_template()`, and keeps the generation arguments wired through to `generate()`.

### CODE REVIEW

**Comment by mitmul** [comment]
Friendly ping — I verified this locally with the new Plamo2VL path and added tests for processor loading, prompt templating, config token IDs, and input embeddings behavior.
There are also two small supporting fixes included for compatibility (gelu_pytorch_tanh handling and the quantization predicate helper used by tests).
I’d appreciate a review when you have a chance, especially on whether you’d prefer the support code split into a separate PR.

**Comment by mitmul** [comment]
@Blaizzy This PR is mergeable and ready for review. If the scope is too broad, I can immediately split out the supporting changes (lockfile / helper refactor / docs) and keep this PR focused on Plamo2VL support only. Would you prefer that, or is this shape acceptable?

**Review by Blaizzy — CHANGES_REQUESTED** [review]
Please look at my other PRs adding models and ajust yours to match or open an issue and I will handle it

**Comment by mitmul** [comment]
> Please look at my other PRs adding models and ajust yours to match or open an issue and I will handle it

Thanks for the review and for pointing me to the recent model-addition PRs. I’ll align this with the recent model-addition PRs rather than opening an issue.

I’ll update the branch against current main, drop unrelated files such as `uv.lock` and `CONTRIBUTING.md`, move the custom processor loading to the model-local `install_auto_processor_patch` pattern, and keep shared changes only where Plamo2VL strictly requires them. I’ll also rerun the focused tests and push the cleaned-up PR.

**Comment by mitmul** [comment]
I’ve pushed the cleanup to align this with the recent model-addition pattern. The PR is now updated against current main, the lockfile/docs/shared loader changes were removed, and the custom processor loading is registered locally via `install_auto_processor_patch`.

Re-requesting review now.

**Comment by mitmul** [comment]
Hi @Blaizzy, sorry for the extra ping. I’ve responded to the review feedback and requested a re-review, but it may have gotten buried in notifications.

When you have a chance, could you please take another look at this PR? I’d really appreciate any guidance on what else is needed to move it forward.

Thanks again for maintaining mlx-vlm!

File: mlx_vlm/models/llava/vision.py [hunk 1]
```diff
@@ -81,12 +82,27 @@ def __call__(self, queries, keys, values, mask=None):
 class MLP(nn.Module):
     def __init__(self, config: VisionConfig):
         super().__init__()
-        self.activation_fn = nn.GELU(approx="fast")
+        self.hidden_act = getattr(config, "hidden_act", "gelu")
         self.fc1 = nn.Linear(config.hidden_size, config.intermediate_size)
         self.fc2 = nn.Linear(config.intermediate_size, config.hidden_size)
 
+    def _gelu_pytorch_tanh(self, x: mx.array) -> mx.array:
```
Comment (lucasnewman) [inline mlx_vlm/models/llava/vision.py]: This is just nn.GELU(approx="precise"), we don't need a separate implementation.

**Comment by lucasnewman** [comment]
@mitmul Overall looking good. The model needs a README in the style of other READMEs in the repo. If you can add that and fix the activation issue I mentioned, we can merge this. Thanks!

**Review by lucasnewman — CHANGES_REQUESTED** [review]
See comments.

File: mlx_vlm/models/llava/vision.py (hunk 1 above)
Comment (mitmul) [reply mlx_vlm/models/llava/vision.py]: Thanks for pointing this out. I removed the separate `_gelu_pytorch_tanh` implementation and updated the LLaVA vision MLP to use `nn.GELU(approx="precise")` for `gelu_pytorch_tanh` / `gelu_new`. This is pushed in 381e2c3.

**Comment by mitmul** [comment]
@lucasnewman Thanks, updated the activation path to use `nn.GELU(approx="precise")` and removed the custom `gelu_pytorch_tanh` implementation.

I also added the Plamo2VL README in the same style as the other model READMEs, updated it to cover both the 8B and 2B variants, and verified the documented CLI/Python examples locally.

While validating the README examples, I found and fixed two small runtime issues: `tokenizer.jsonl` was not included in the Hugging Face snapshot download allowlist, and the Plamo2VL language wrapper needed to tolerate extra generation kwargs from the CLI. I added focused regression tests for both cases and reran the Plamo2VL-focused test set successfully.

**Comment by mitmul** [comment]
Hi @lucasnewman, gentle follow-up on this PR.

I’ve addressed the items from your last review: added the Plamo2VL README in the style of the existing model READMEs, updated the activation path to use `nn.GELU(approx="precise")`, and added focused regression tests for the small runtime issues found while validating the examples.

When you have a chance, could you please let me know if there is anything else you’d like changed before this can move forward?

Thanks again to you and @Blaizzy for maintaining mlx-vlm. I really appreciate the work you put into this project.

**Comment by lucasnewman** [comment]
@mitmul Thanks, it's looking good. Can you run `pre-commit run --all` to clear the tests and then I'll merge?

**Comment by mitmul** [comment]
@lucasnewman Done, thank you. I ran `pre-commit run --all`; the first run applied the expected black/isort formatting changes, and the second run passed cleanly.

I pushed the formatting-only commit as 9b673e0.

The new GitHub Actions run for this commit currently shows `action_required` with no jobs, so it may need maintainer approval before it can run on GitHub.

**Comment by mitmul** [comment]
@lucasnewman Hi, sorry to ping again, and thank you very much for all the work you put into maintaining this excellent project.

I have updated this PR with the latest `upstream/main` and resolved the merge conflicts. I also reran the focused tests around the resolved areas, including the CLI generation smoke/video forwarding tests and the Plamo2VL/moondream2 model tests, and they are passing locally.

I fully understand that review bandwidth is limited, but when you have a chance, I would be very grateful if you could take another look or let me know if there is anything else I should adjust. Thank you again for your time and for maintaining this repository.

**Comment by mitmul** [comment]
Hi @lucasnewman , apologies for pinging you again. I’ve resolved the latest conflicts with `main` and rerun `pre-commit run --all`; all checks passed.

Thank you sincerely for the time and care you put into maintaining this excellent project, and for reviewing this contribution. I greatly appreciate that maintainer time is limited, so there is absolutely no urgency. Whenever it is convenient, I would be very grateful if you could take another look at this PR.

Please let me know if there is anything else I should revise. Thank you again!

**Review by lucasnewman — APPROVED** [review]

**Comment by lucasnewman** [comment]
@mitmul Can you [sign](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) your commits and force push so we can clear the merge requirements?

**Comment by mitmul** [comment]
@lucasnewman Done — I’ve signed the commits and force-pushed the updated branch. Could you please check whether the merge requirements are now cleared? Thanks!

**Review by lucasnewman — APPROVED** [review]

**Comment by lucasnewman** [comment]
@Blaizzy Can you clear your changes requested so we can merge?

**Comment by mitmul** [comment]
Hi @lucasnewman, sorry to ping you again. While the remaining change-request review from @Blaizzy was pending, `main` moved forward and the PR developed merge conflicts. I’ve now resolved the conflicts, updated the branch, and rerun the relevant tests and pre-commit checks successfully.
When you have a moment, could you please take another look? No rush at all, and thank you again for your time and help with this PR.

**Review by lucasnewman — APPROVED** [review]

**Comment by Blaizzy** [comment]
LGTM, thanks @mitmul!

**Review by Blaizzy — APPROVED** [review]
